### PR TITLE
document rerun failed tests with knapsack pro

### PR DESCRIPTION
**Pull request by 3v0k4**
# Description
This PR adds a section to https://circleci.com/docs/rerun-failed-tests/#additional-examples on how to configure Knapsack Pro to use the CircleCI rerun failed tests feature.

# Reasons
A DevOps Customer Engineer from CircleCI (Denis) signed up for Knapsack Pro to document the integration himself after some customers asked CircleCI how to integrate the two.

Knapsack Pro already has customers using the rerun failed tests feature, so we decided to take over the task and contribute to the docs.

Notice that Knapsack Pro is already mentioned in CircleCI docs on https://circleci.com/docs/parallelism-faster-jobs/#other-ways-to-split-tests

# Content Checklist
Please follow our style when contributing to CircleCI docs. Our style guide is here: [https://circleci.com/docs/style/style-guide-overview](https://circleci.com/docs/style/style-guide-overview).

Please take a moment to check through the following items when submitting your PR (this is just a guide so will not be relevant for all PRs) 😸:

- [x] Break up walls of text by adding paragraph breaks.
- [x] Consider if the content could benefit from more structure, such as lists or tables, to make it easier to consume.
- [x] Keep the title between 20 and 70 characters.
- [x] Consider whether the content would benefit from more subsections (h2-h6 headings) to make it easier to consume.
- [x] Check all headings h1-h6 are in sentence case (only first letter is capitalized).
- [x] Is there a "Next steps" section at the end of the page giving the reader a clear path to what to read next?
- [x] Include relevant backlinks to other CircleCI docs/pages.

### CODE REVIEW

**Comment by 3v0k4** [comment]
I also added the docs for Minitest + Knapsack Pro in a separated commit.

Let me know if there's anything I can help with in this PR 🙏

File: jekyll/_cci2/rerun-failed-tests.adoc [hunk 1]
```diff
@@ -116,6 +116,37 @@ The job should only rerun tests that are from a `classname` of `file` that had a
 [#additional-examples]
 == Additional examples
 
+[#configure-a-job-running-ruby-minitest-tests-across-parallel-ci-nodes-with-knapsack-pro]
+=== Configure a job running Ruby (minitest) tests across parallel CI nodes with Knapsack Pro
+
+Knapsack Pro dynamically distributes your tests based on up-to-date test execution data to achieve the perfect split.
+
+To use Knapsack Pro with the CircleCI rerun failed tests feature, you just need to:
```
Comment (rosieyohannan) [inline jekyll/_cci2/rerun-failed-tests.adoc:124]:
```suggestion
To use Knapsack Pro with the CircleCI rerun failed tests feature, follow these steps:
```

File: jekyll/_cci2/rerun-failed-tests.adoc [hunk 2]
```diff
@@ -141,6 +172,47 @@ gem 'rspec_junit_formatter'
 . Ensure you are using `xargs` in your `circleci tests run` command to pass the list of test files/classnames via stdin to `--command`.
 . Update the `glob` command to match your use case. See the RSpec section in the xref:collect-test-data#rspec[Collect Test Data] document for details on how to output test results in an acceptable format for `rspec`. **If your current job is using xref:test-splitting-tutorial#[CircleCI's intelligent test splitting], you must change the `circleci tests split` command to `circleci tests run` with the `--split-by=timings` parameter.** If you are not using test splitting, `--split-by=timings` can be omitted.
 
+[#configure-a-job-running-ruby-rspec-tests-across-parallel-ci-nodes-with-knapsack-pro]
+=== Configure a job running Ruby (rspec) tests across parallel CI nodes with Knapsack Pro
+
+Knapsack Pro dynamically distributes your tests based on up-to-date test execution data to achieve the perfect split.
+
+To use Knapsack Pro with the CircleCI rerun failed tests feature, you just need to:
```
Comment (rosieyohannan) [inline jekyll/_cci2/rerun-failed-tests.adoc:180]:
```suggestion
To use Knapsack Pro with the CircleCI rerun failed tests feature, follow these steps:
```

File: jekyll/_cci2/rerun-failed-tests.adoc [hunk 3]
```diff
@@ -141,6 +172,47 @@ gem 'rspec_junit_formatter'
...
+
+ - store_test_results:
+     path: /tmp/test-results
+
+ - store_artifacts:
+     path: /tmp/test-results
+     destination: test-results
+```
+
+. You may also enable link:https://docs.knapsackpro.com/ruby/split-by-test-examples/[Split by Test Examples] to parallelize tests across CI nodes by individual `it`/`specify`. This is useful when you have slow test files but don't want to manually split test examples into smaller test files.
```
Comment (rosieyohannan) [inline jekyll/_cci2/rerun-failed-tests.adoc:214]:
```suggestion
. You may also enable link:https://docs.knapsackpro.com/ruby/split-by-test-examples/[Split by Test Examples] to parallelize tests across CI nodes by individual `it`/`specify`. This is useful when you have slow test files but do not want to manually split test examples into smaller test files.
```

**Review by rosieyohannan — COMMENTED** [review]
Thank you! Just a couple of minor changes then we can get this published 🎉

**Comment by 3v0k4** [comment]
@rosieyohannan I committed your suggestions. Thanks! 🙏

Please let me know if I can help with anything else.

**Review by rosieyohannan — APPROVED** [review]

**Comment by sebastian-lerner** [comment]
hey @ArturT and @3v0k4, I'm helping someone out who's trying to use these docs to set up "rerun failed tests" with knapsack.

The snippet is below which looks like it should do what we want. but on the failed test re-runs, it's still running all of the tests. I'm seeing in the logs that CircleCI recognizes that there were only 3 test files to rerun, but knapsack is wanting to run all of the test files instead of those 3.  Anything stick out that might be causing that?  

```
export KNAPSACK_PRO_TEST_FILE_LIST_SOURCE_FILE=~/docfox-rework/tmp/system_tests_to_run.txt
shopt -s globstar
circleci tests glob test/**/*_test.rb | circleci tests run --index 0 --total 1 --command ">$KNAPSACK_PRO_TEST_FILE_LIST_SOURCE_FILE xargs -n1 echo" --verbose
KNAPSACK_PRO_TEST_DIR=test KNAPSACK_PRO_FIXED_QUEUE_SPLIT=true RAILS_ENV=test ./bin/rake "knapsack_pro:queue:minitest[--no-focus --verbose --ci-report --no-ci-clean]
```

**Comment by ArturT** [comment]
@sebastian-lerner Please remove [`KNAPSACK_PRO_FIXED_QUEUE_SPLIT=true`](https://docs.knapsackpro.com/ruby/reference/#knapsack_pro_fixed_queue_split-queue-mode). Then the default value `false` is used for the CircleCI. This should resolve the issue. 

You could also refer the customer to our support email and we would be happy to directly help.
https://knapsackpro.com/contact

**Comment by sebastian-lerner** [comment]
aha! thank you! i'll pass that information along.

**Pull request merged**